### DISCUSSION TITLE
About the example "4.2 Create an egs++ input file", i really don't understand why some programming mean?

### DISCUSSION THREAD

**Discussion by kstcmcjw** [nrc-cnrc/EGSnrc#820] [Questions]
Hello each elder of EGSnrc,  About the example "4.2 Create an egs++ input file", i really don't understand why some programming mean. There are some problom pozzled me.
1. Why Z-planes has four numbers?  X  and Y position only has two? Three-dimensional graphics depend on the beginning and end of each direction to draw three-dimensional graphics, Why are there four points on the Z-Plane?
![image](https://user-images.githubusercontent.com/96934217/148628018-14085027-04af-4c9b-b240-5e2f98c627d5.png)

2. As we have define the energy of photons or electrons in 20MeV , why the "Media" has the cutoff energies?  I think the maximum energy is 20MeV
![image](https://user-images.githubusercontent.com/96934217/148628128-fe9a5cda-50fb-43e7-a00d-43a5964b7dd0.png)

3. And the figure only see the Ta plate , lack we have define the lead and the water's simulation?
 
![image](https://user-images.githubusercontent.com/96934217/148628359-b9efd605-26b2-4b99-b1b5-f58fbe82f809.png)
![image](https://user-images.githubusercontent.com/96934217/148628366-2685edf8-1f31-4dc4-9e89-e5dae5599e33.png)
![image](https://user-images.githubusercontent.com/96934217/148633211-ee5f1bd2-e091-47e0-9336-91eb7fa4cd4a.png)

i can't find where the "lead" the "water" is?  and there is no lead and water in this figure!

4. The EGSnrc program’s execute interface after start, execution failed, and there is NO batch, why? i was really puzzled!

**my slab.egsinp programming is:**

```
###1.Run control
:start run control:
    ncase = 1e3
:stop run control:

###2.Geometry
:start geometry definition:
    :start geometry:
        name = slab
        library = egs_ndgeometry
        type = EGS_XYZGeometry
        x-planes = -5, 5 
        y-planes = -5, 5  
        z-planes = -10, 0, 0.1, 10
        :start media input:
            media = vacuum tantalum
            set medium = 1 1
        :stop media input:
    :stop geometry:
    simulation geometry = slab
:stop geometry definition:

###3.Media
:start media definition:
    ae = 0.521  
    ap = 0.01                   
    ue = 50.511                                
    up = 50                               
   
    :start tantalum:
          density correction file = tantalum
    :stop tantalum:

    :start lead:
          density correction file = lead
    :stop lead:

    :start water:
          density correction file = water_liquid
    :stop water:

:stop media definition:

###4.Source
:start source definition:

    pencil beam
    :start source:
        name = pencil_beam
        library = egs_parallel_beam
        charge = -1
        direction = 0 0 1
        :start spectrum:
            type = monoenergetic
            energy = 20                         
        :stop spectrum:
        :start shape:
            type = point
            position = 0 0 -10                  
        :stop shape:
    :stop source:
    simulation source = pencil_beam

:stop source definition:

###5.VIEWER CONTROL
:start view control:
    set color = lead 120 120 200 200
    set color = tantalum 120 255 120 255
    set color = water 0 220 255 200
:stop view control:

###6.AUSGAB OBJECTS
:start ausgab object definition:
    :start ausgab object:
        name = tracks
        library = egs_track_scoring
    :stop ausgab object: 
:stop ausgab object definition:
```

![image](https://user-images.githubusercontent.com/96934217/148633802-99d99db7-851e-4da8-8521-e06da869fc74.png)

Thanks for your attention and answers!

**Comment by rtownson** [comment]
1. The 4 number define 3 regions, or 3 slabs where you can define a material. Between -10 and 0 is vacuum, between 0 and 0.1 is tantalum, and between 0.1 and 10 is vacuum. If you only defined between 0 and 0.1, then you would not see particle tracks anywhere outside the tantalum track, because the space is not defined! Particles outside any defined regions are discarded.
2. The media have independent cut-off energies because you may wish to change them for some more advanced studies. There is no reason to have them larger than your maximum energy in the simulation, so you could indeed reduce them to 20.511 and 20 MeV. However, there is little harm to having them a bit too large, the main effect is that it expands the interpolation grid (so don't have AE and AP at 100 MeV for simulations at 100keV). 
3. It is possible to define materials that are not used in the simulation. In this case, yes there are some extra materials defined and not used. A material is only included in the simulation if it is actually assigned to a geometry region, as is the case for tantalum (where you see `set medium = 1 1`.
4. I copy/pasted your input file and it ran fine using egs_app for me. Can you try with egs_app instead of myapp? You may have to compile the application first. If it still fails, copy/paste all the output so I can look for other errors. Have you have any simulations run successfully yet? Try the existing example slab.egsinp in egs_home/egs_app.

**Reply by kstcmcjw** [reply]
> Collaborator
thank you for your answer!
1. Yes, your are right! As you mentioned i have a try of two numbers, instead of four numbers, the new slab.ptracks file have generated is more than 900MB, and the simulation figure did not show the particle way after collision, as the below shows:
![image](https://user-images.githubusercontent.com/96934217/148680568-9ab57313-e34c-4023-8d41-6af5056e0a36.png)
![image](https://user-images.githubusercontent.com/96934217/148680599-121db62c-41ee-487d-8434-4ccf2049ec8e.png)
2. A very vivid explain! I see, if i set a very big range of energy, the interpolation grid method may not accuracy.
3. Thank you, i will have a try with other materials.
4. the fail message is as the below:, but the figure is right, why?

```
In EGS_Application::initSimulation()
**********************************************
EGS_RandomGenerator::createRNG: no 'RNG definition' input
EGS_Application::initRNG(): using default RNG
================================================================================
EGSnrc version 4 for win3264                           Sun Jan  9 20:21:56 2022
================================================================================
configuration....................................................win3264
user code........................................................myapp
pegs file........................................................
using host.......................................................BF-202009142050
input file.......................................................slab
output file(s)...................................................slab
================================================================================
In initCrossSections(): spin effects = 1
 ===> Photonuclear flag:   0
 RAYLEIGH OPTION REQUESTED FOR MEDIUM NUMBER  1

  PEGSLESS INPUT.  CALCULATING ELECTRON CROSS-SECTIONS.
  density_file tantalum                                                                                                                                                                                                                                                        

  Medium data: 

 AE =    5.21000E-01 MeV,  UE =    5.05110E+01 MeV
 AP =    1.00000E-02 MeV,  UP =    5.00000E+01 MeV

  No material data file supplied.  Material data obtained from
  .egsinp file or density correction file.

   Medium: tantalum                
 Sterncid: tantalum                
     rho:    1.66540E+01 g/cm^3
 Elements: TA  
    rhoz:    1.00000E+00
      pz:    5.52645E-03
  iunrst:     0
  iaprim:     0
    gasp:    0.00000E+00 atm.
  density correction file: C:\EGSnrc\HEN_HOUSE\pegs4\density_corrections\elements\tantalum.density


(Re)-initializing photon cross sections with files from the series: xcom

 Compton cross sections: default
 Using Compton cross sections from C:\EGSnrc\HEN_HOUSE\data\compton_sigma.data
 Working on medium   1 ... 
  ->  100 atomic ff values computed!
      preparing data for Rayleigh sampling ... done

OK

Reading screened Rutherford MS data ...............  done 


Reading spin data base from C:\EGSnrc\HEN_HOUSE\data\spinms.data
EGSnrc spin data, version 2.0   
Data generated on a machine with 1234 endianess
The endianess of this CPU is 1234
Ranges:      1.00   100.00  0.30054  1.00000


  medium    1 .....................  done
  
  Medium            1  sige =    10.311986840371722        9.9678018166723330       monotone =  F F
  
  Initializing tmxs for estepe =   0.25000000000000000       and ximax =   0.50000000000000000     
  

Bound Compton scattering requested, reading data ......  Done

 Initializing Bound Compton scattering ......
 Medium   1 has  21 shells: 
   1  861   1  0.02740 0.141E+01    67.416
   2  862   2  0.02740 0.503E+01    11.681
   3  863   3  0.02740 0.247E+01    11.136
   4  864   4  0.05479 0.271E+01     9.881
   5  865   5  0.02740 0.111E+02     2.708
   6  866   6  0.02740 0.652E+01     2.469
   7  867   7  0.05479 0.696E+01     2.194
   8  868   8  0.05479 0.452E+01     1.793
   9  869   9  0.08219 0.462E+01     1.735
  10  870  10  0.02740 0.234E+02     0.563
  11  871  11  0.02740 0.148E+02     0.462
  12  872  12  0.05479 0.158E+02     0.402
  13  873  13  0.05479 0.121E+02     0.239
  14  874  14  0.08219 0.124E+02     0.227
  15  875  15  0.02740 0.607E+02     0.068
  16  876  16  0.02740 0.414E+02     0.042
  17  877  17  0.05479 0.444E+02     0.034
  18  878  18  0.08219 0.117E+02     0.025
  19  879  19  0.10959 0.118E+02     0.023
  20  880  20  0.05479 0.944E+02     0.008
  21  881  21  0.01370 0.221E+03     0.008
 ...... Done.

Output from subroutine EDGSET:
==============================
 Atomic relaxations requested! 

 Reading simplified photo-absorption data ..... Done

 Reading simplified relaxation data ..... Done

 Reading parametrized XCOM photo cross section data ..... Done

 Reading EADL relaxation data ......
  Z =  73 has 22 shells
 ...... Done.

 EGSnrc SUCCESSFULLY 'HATCHED' FOR ONE MEDIUM.


The following media are defined:
================================

  0  tantalum                 AE= 0.5210 AP= 0.0100 0


Transport parameter and cross section options:
==============================================
Photon cross sections                             xcom            
Compton cross sections                            default         
Global Pcut                                       0.01
Pair cross sections                               BH
Pair angular sampling                             Simple
Triplet production                                Off
Bound Compton scattering                          norej
Radiative Compton corrections                     Off
Rayleigh scattering                               On
Atomic relaxations                                On
Photoelectron angular sampling                    On
Photonuclear attenuation                          Off
Photonuclear cross sections                       default         

Global Ecut                                       0.521
Brems cross sections                              BH
Brems angular sampling                            KM
Spin effects                                      On
Electron Impact Ionization                        Off             
Global Smax                                       1e+010
ESTEPE                                            0.25
Ximax                                             0.5
Boundary crossing algorithm                       Exact
Skin depth for BCA                                3
Electron-step algorithm                           EGSnrc
==============================================



======================== geometry =====================
 type = EGS_XYZGeometry
 name = slab
 number of regions = 3
======================== geometry =====================
 type = x-planes
 name = geometry0
 number of regions = 1
 first plane at -5, 1 slices of 10 thickness
=======================================================
======================== geometry =====================
 type = y-planes
 name = geometry1
 number of regions = 1
 first plane at -5, 1 slices of 10 thickness
=======================================================
======================== geometry =====================
 type = z-planes
 name = geometry2
 number of regions = 3
 plane positions = -10 0 0.1 10 
=======================================================
=======================================================


The simulation uses the following source:
=========================================
 Parallel beam from a shape of type point with monoenergetic 20 MeV, electrons


Random number generator:
============================================
  type                = ranmar
  high resolution     = no
  initial seeds       = 33 97
  numbers used so far = 0


Run Control Object (RCO):
=========================
  type = simple


The following ausgab objects are included in the simulation
===========================================================


Particle Track Scoring (tracks)
======================================================
 - Scoring photon tracks       = YES
 - Scoring electron tracks     = YES
 - Scoring positron tracks     = YES
 - First event to score        = 0
 - Last event to score         = 1024
 - Track buffer size           = 1024
 - Output file name            = C:\EGSnrc\EGS_HOME\myapp\slab.ptracks


*******************************************
Dose Scoring Object (dose)
*******************************************

 - Regions in dose calculator :1


--------------------------------------
medium      rho/[g/cm**3]
--------------------------------------
tantalum 16.65
--------------------------------------

*******************************************





Starting simulation on Sun Jan 09 20:21:56 2022

    Fresh simulation of 1000 histories



Running 1000 histories

  Batch             CPU time        Result   Uncertainty(%)
==========================================================
      1


*********** execution failed *********
```

**Reply by rtownson** [reply]
I think the simulation is failing because the source exists outside the defined geometry of the simulation. The tracks file is probably an old one from the previous simulation - try deleting it and I think you will find the simulation is not creating a new one.

**Reply by kstcmcjw** [reply]
Thank you for your attantion, i am sure i delete the created files when simulation failed, as your said "the source exists outside the defined geometry", i have examed the define:

```
###2.Gemotry
......
type = EGS_XYZGeometry
x-planes = -5, 5
y-planes = -5, 5
z-planes = -10, 0, 0.1, 10
:start media input:
media = vacuum tantalum
......

###4.Source
pencil beam
:start source:
    name = pencil_beam
    library = egs_parallel_beam
    charge = -1
    direction = 0 0 1
```

The site 0 0 1 is include in the space of (-5,5) (-5,5) (-10,10),

and there are one doubt that seems i have mistake the program?
![image](https://user-images.githubusercontent.com/96934217/148862996-b5d04e05-dbfc-4c7f-80bd-536d7df4228a.png)

use cmd.exe i didn't get the  c:\EGSnrc\EGS_HOME\myapp\   -i slab.egsinp 
is that problem?   
Thank for your answer!

**Reply by rtownson** [reply]
When you change back to `-10, 0, 0.1, 10` does the simulation still crash? What has changed since the simulation ran correctly?

You will not see a file `myapp\ -i slab.egsinp`. This is just telling you how to launch the simulation, which it seems like you were already doing. E.g. you navigate to `c:\EGSnrc\EGS_HOME\myapp\` using `cd`, and then launch the simulation with `myapp -i slab.egsinp`.

**Reply by kstcmcjw** [reply]
I am sure the z-plane is：z-planes = -10, 0, 0.1, 10， my question is why the slab_ptracks is OK, but the execution filed?
![image](https://user-images.githubusercontent.com/96934217/148958978-527c0ff3-c0c2-4f1b-9cf4-333c1ce33712.png)
![image](https://user-images.githubusercontent.com/96934217/148959079-0271f1bc-1c7b-4410-a850-56f5e3a6d65d.png)
![image](https://user-images.githubusercontent.com/96934217/148959099-c3d6ff3e-a69e-4ccb-95e2-5222ac19c302.png)
and i change the history into 100 counts
![image](https://user-images.githubusercontent.com/96934217/148959218-9bac5b27-b7ab-4c8c-a3a3-afd02610af37.png)

and as you see, why the figure is right, but the reslut is failed? It seems that the program will calculation but an error or a special incident interrupt it?

**Reply by rtownson** [reply]
I see, I had forgotten that you saw this error from the start. Did you try with egs_app instead of myapp like I suggested? I'm not sure why you are seeing this error. Do you get any additional information with the error when you launch the simulation from the command line instead of using egs_gui?

**Reply by kstcmcjw** [reply]
Yesterday, i had a try of egs_app but failed, OK, i will have a try once more, and make printscreen step by step, by the way, your said egs_app is in this folder?
![image](https://user-images.githubusercontent.com/96934217/148968554-fd894735-824a-44ae-95b6-ae2fa182ab6b.png)
This is the running reslut:
![image](https://user-images.githubusercontent.com/96934217/148969583-921f9416-0ca4-4689-aee1-6c60798451c1.png)
![image](https://user-images.githubusercontent.com/96934217/148969658-e38ff674-67c7-44ec-bc35-99c931971461.png)

Is my software will re-install? Or have something wrong?

**Reply by rtownson** [reply]
Yes that's the right folder. Did you compile egs_app first?

**Reply by kstcmcjw** [reply]
i first use the EGS_gui to run it, and as you see, whether i should use the cmd window?
![image](https://user-images.githubusercontent.com/96934217/148970695-199d7f62-0f80-4040-aec0-e3fc95b59d9f.png)
![image](https://user-images.githubusercontent.com/96934217/148970719-c97a11d9-c514-4a0a-a3a6-46debe2ed3e3.png)

**Reply by rtownson** [reply]
Since it seems like the simulation is actually running correctly, I would try running it using the command window to see if it gives you any extra output. Sometimes it will show slightly more information.

**Comment by kstcmcjw** [comment]
And i have another question: as we know, over 10MeV photon or electron have a collision with material may cause neutron, duo to photonuclear reaction. But in our experiment, simulation energy to 20MeV, but there seems no neutron created or mentioned, why? Whether the EGS software only simulation the elecrton, positron and photon except the other particles?

**Reply by rtownson** [reply]
EGSnrc can model photonuclear attenuation if you set the transport parameter: `Photonuclear attenuation = On`. By default it is off. However, EGSnrc does not actually simulate the transport of neutrons; only photons, electrons and positrons. So you will not see neutrons in the particle tracks.
